client/src/websockets.rs

```diff
@@ -0,0 +1,632 @@
+use std::time::{Duration, Instant, SystemTime};
+use surf::http::headers::ToHeaderValues;
+use tokio::net::TcpStream;
+use tokio::runtime::Handle;
+use tokio::sync::Mutex;
+use tokio::task::{yield_now, AbortHandle, JoinHandle, Unconstrained};
+use tokio::time::{sleep, timeout};
+use tokio::*;
+use tokio_native_tls::TlsConnector;
+use tokio_tungstenite::tungstenite::client::IntoClientRequest;
+use tokio_tungstenite::tungstenite::protocol::{frame, WebSocketConfig};
+use tokio_tungstenite::tungstenite::{http, WebSocket};
+use tokio_tungstenite::*;
+use tokio_tungstenite::{connect_async, tungstenite::protocol::Message};
+use url::Url;
+
+//Constants used by Signal, and therefore also used in our project
+const SECOND: u32 = 1000;
+const MINUTE: u32 = SECOND * 60;
+const HOUR: u32 = MINUTE * 60;
+const DAY: u32 = HOUR * 24;
+const WEEK: u32 = DAY * 7;
+const MONTH: u32 = DAY * 30;
+// 30 seconds + 5 seconds for closing the socket above.
+const KEEPALIVE_INTERVAL_MS: u32 = 30 * SECOND;
+
+// If the machine was in suspended mode for more than 5 minutes - trigger
+// immediate disconnect.
+const STALE_THRESHOLD_MS: u32 = 5 * MINUTE;
+
+// If we don't receive a response to keepalive request within 30 seconds -
+// close the socket.
+const KEEPALIVE_TIMEOUT_MS: u32 = 30 * SECOND;
+
+const MAX_MESSAGE_SIZE: u32 = 512 * 1024;
+
+struct StreamHandler {
+    pub stream: WebSocketStream<tokio_native_tls::TlsStream<TcpStream>>,
+}
+
+impl StreamHandler {
+    async fn try_new() -> Result<StreamHandler> {
+        let stream = open_ws_connection_to_server_as_client().await?;
+        Ok(StreamHandler { stream })
+    }
+}
+
+#[derive(Clone)]
+pub(crate) struct WebsocketHandler {
+    socket: Arc<Mutex<StreamHandler>>,
+    receiver_handle: Arc<thread::JoinHandle<JoinHandle<()>>>,
+    text_channel: Arc<Mutex<mpsc::Receiver<Message>>>,
+    ws_request_channel: Arc<Mutex<mpsc::Receiver<WebSocketRequestMessage>>>,
+    outgoing_id: Arc<Mutex<u64>>,
+    outgoing_map: Arc<Mutex<HashMap<u64, AbortHandle>>>,
+    incoming_response: Arc<Mutex<HashMap<u64, WebSocketResponseMessage>>>,
+    keepalive_handler: Option<Arc<Mutex<KeepaliveHandler>>>,
+    has_keepalive: bool,
+}
+impl WebsocketHandler {
+    pub async fn try_new(keepalive_options: Option<KeepAliveOptions>) -> Result<WebsocketHandler> {
+        let socket = Arc::new(Mutex::new(StreamHandler::try_new().await?));
+        let res: Arc<Mutex<HashMap<u64, WebSocketResponseMessage>>> =
+            Arc::new(Mutex::new(HashMap::new()));
+        let out: Arc<Mutex<HashMap<u64, AbortHandle>>> = Arc::new(Mutex::new(HashMap::new()));
+        let (handle, text_channel, request_channel) =
+            WebsocketHandler::get_handle_for_ws_listener(socket.clone(), res.clone(), out.clone());
+
+        let mut ws_handler = WebsocketHandler {
+            socket,
+            receiver_handle: Arc::new(handle),
+            text_channel: Arc::new(Mutex::new(text_channel)),
+            outgoing_id: Arc::new(Mutex::new(1)),
```

Comment:
Hvad er grunden til `1`, skal det læses om init værdien?

Reply:
ja

Reply:
Signal-Desktop starter med id=1 og det er er i en arc mutex for at sikre id altid stiger, selv hvis WebsocketHandler bliver cloned